cranelift/assembler-x64/meta/src/dsl/encoding.rs

```diff
@@ -705,98 +751,309 @@ impl fmt::Display for OpcodeMod {
-    _OF38,
+/// Contains allowed VEX length definitions.
+///
+/// VEX encodes these in the `L` bit field, a single bit with `128-bit = 0` and
+/// `256-bit = 1`. For convenience, we also include the `LIG` and `LZ`
+/// syntax, used by the reference manual, and always set these to `0`.
+pub enum VexLength {
+    /// Set `VEX.L` to `0` (128-bit).
+    L128,
+    /// Set `VEX.L` to `1` (256-bit).
+    L256,
+    /// Set `VEX.L` to `0`, but not necessarily for 128-bit operation. From the
+    /// reference manual: "The VEX.L must be encoded to be 0B, an #UD occurs if
+    /// VEX.L is not zero."
+    LZ,
+    /// The `VEX.L` bit is ignored (e.g., for floating point scalar
+    /// instructions). This assembler will emit `0`.
+    LIG,
 }
 
-impl fmt::Display for VexMMMMM {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+impl VexLength {
+    /// Encode the `L` bit.
+    pub fn bits(&self) -> u8 {
         match self {
-            VexMMMMM::None => write!(f, "None"),
-            VexMMMMM::_OF => write!(f, "_0F"),
-            VexMMMMM::_OF3A => write!(f, "_OF3A"),
-            VexMMMMM::_OF38 => write!(f, "_OF38"),
+            Self::L128 | Self::LIG | Self::LZ => 0b0,
+            Self::L256 => 0b1,
         }
     }
 }
 
-/// Describe the register index to use. This wrapper is a type-safe way to pass
-/// around the registers defined in `inst/regs.rs`.
-#[derive(Debug, Copy, Clone, Default)]
-pub struct Register(u8);
-impl From<u8> for Register {
-    fn from(reg: u8) -> Self {
-        debug_assert!(reg < 16);
-        Self(reg)
+impl fmt::Display for VexLength {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Self::L128 => write!(f, "128"),
+            Self::L256 => write!(f, "256"),
+            Self::LIG => write!(f, "LIG"),
+            Self::LZ => write!(f, "LZ"),
+        }
     }
 }
-impl From<Register> for u8 {
-    fn from(reg: Register) -> u8 {
-        reg.0
-    }
+
+/// The VEX encoding, introduced for AVX instructions.
+///
+/// ```
+/// # use cranelift_assembler_x64_meta::dsl::{vex, VexLength::L128};
+/// // To encode a BLENDPD instruction in the manual: VEX.128.66.0F3A.WIG 0D /r ib
+/// let enc = vex(L128)._66()._0f3a().wig().op(0x0D).r().ib();
+/// assert_eq!(enc.to_string(), "VEX.128.66.0F3A.WIG 0x0D /r ib");
+/// ```
+pub struct Vex {
+    /// The length of the operand (e.g., 128-bit or 256-bit).
+    pub length: VexLength,
+    /// Map the `PP` field encodings.
+    pub pp: Option<VexPrefix>,
+    /// Map the `MMMMM` field encodings.
+    pub mmmmm: Option<VexEscape>,
+    /// The `W` bit.
+    ///
+    /// Since VEX-encoded instructions may ignore the `W` bit, we model this
+    /// with a three-state field:
+    /// - `None`: the `W` bit is ignored (i.e., `WIG` in the manual)
+    /// - `Some(false)`: the `W` bit is set to `0` (i.e., `W0` in the manual)
+    /// - `Some(true)`: the `W` bit is set to `1` (i.e., `W1` in the manual)
+    pub w: Option<bool>,
```

Comment:
Could we define a `WBit` (or pick your favorite name) enum here with `WIG`/`W0`/`W1`?